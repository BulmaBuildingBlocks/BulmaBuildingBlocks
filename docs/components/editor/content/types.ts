export enum EditableContentTypes {
  Text = 'text',
  Title = 'title'
}

export interface EditableContentProps {
  /**
   * Element wrapping the Editable Content.
   */
  tag?: string;

  /**
   * Provides different editing options
   */
  type?: EditableContentTypes;

  /**
   * Class placed on the parent element.
   */
  class?: string;

  /**
   * Value that's edited.
   */
  value: string;

  /**
   * Content Editable.
   */
  editable?: boolean;
}
