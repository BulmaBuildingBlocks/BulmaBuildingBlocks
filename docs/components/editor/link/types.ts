import { Status } from '~/shared/config';

export enum LinkType {
  Link = 'Link',
  Button = 'button',
  Rounded = 'button is-rounded',
  Outlined = 'button is-outlined',
  'Rounded Outlined' = 'button is-rounded is-outlined'
}

export class EditableLink {
  constructor(props: EditableLinkProps) {
    this.label = props.label;
    this.status = props.status || Status.None;
    this.linkType = props.linkType;
    this.href = props.href;
    this.popupPlacement = props.popupPlacement;
  }

  /*
   * Label on the button.
   */
  label: string;

  /**
   * Assigned hyperlink.
   */
  href?: string;

  /**
   * Link styles.
   */
  linkType: LinkType;

  /**
   * Link styles.
   */
  status: Status;

  /**
   * Popup Placement.
   */
  popupPlacement?: string;
}

export interface EditableLinkProps {
  /**
   * Label on the button.
   */
  label: string;

  /**
   * Assigned hyperlink.
   */
  href?: string;

  /**
   * Link styles.
   */
  linkType: LinkType;

  /**
   * Link styles.
   */
  status?: Status;

  /**
   * Popup Placement.
   */
  popupPlacement?: string;
}
