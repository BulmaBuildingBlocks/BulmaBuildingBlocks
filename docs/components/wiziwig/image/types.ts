export interface EditableImageProps {
  /**
   * Key of a permalink referencing the source of the picture.
   */
  src: string;

  /**
   * Caption on the picture, used also as alternative text.
   */
  alt: string;

  /**
   * Hyperlink attached to the picture.
   */
  href?: string;

  /**
   * Picture width.
   */
  width?: number;

  /**
   * Picture height.
   */
  height?: number;

  /**
   * Popup Placement.
   */
  popupPlacement?: string;
}
