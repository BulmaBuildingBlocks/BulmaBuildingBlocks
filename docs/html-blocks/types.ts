import { VueConstructor } from 'vue';

export enum ContentTypes {
  Headers = 'headers',
  Features = 'features',
  CallToActions = 'calltoactions',
  Contents = 'contents',
  Footers = 'footers'
}

export interface Block {
  title: string;
  type?: ContentTypes;
  componentName: string;
  component: VueConstructor;
  content?: any;
  color?: string;
  container?: boolean;
}

export interface BlockListObject {
  [x: string]: VueConstructor;
}
