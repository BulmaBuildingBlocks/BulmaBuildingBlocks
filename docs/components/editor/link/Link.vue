<script lang="ts">
import { CreateElement } from 'vue';
import { Component, Prop, PropSync, Vue } from 'nuxt-property-decorator';
import ButtonToolbar from './LinkToolbar.vue';
import PopupModal from '~/components/PopupModal.vue';

import { EditableLinkProps, LinkType } from '~/components/editor/link/types';
import ClickOutside from '~/directives/click-outside';

/***
 * Editting Features:
 * Upload and replace link
 * Alt tag
 * Link
 */

@Component({
  directives: {
    ClickOutside
  }
})
export default class Link extends Vue {
  @Prop(Boolean) editable!: boolean;
  @PropSync('value', { type: Object }) newValue!: EditableLinkProps;

  modalOpen = false;

  get linkClasses() {
    const linkType = this.newValue.linkType.toLowerCase();
    let linkStyle = '';

    if (this.newValue.status) {
      if (this.newValue.linkType === LinkType.Link) {
        linkStyle = `has-text-${this.newValue.status.toLowerCase()}`;
      } else {
        linkStyle = `is-${this.newValue.status?.toLowerCase()}`;
      }
    }

    return [linkType, linkStyle];
  }

  toggleModal(event: Event) {
    event.preventDefault();

    this.modalOpen = !this.modalOpen;
  }

  closeModal() {
    this.modalOpen = false;
  }

  render(createElement: CreateElement) {
    const link = createElement(
      'a',
      {
        class: this.linkClasses,
        on: {
          click: this.toggleModal
        },
        attrs: {
          href: this.newValue.href
        }
      },
      this.newValue.label
    );

    const linkToolbar = createElement(ButtonToolbar, {
      props: {
        value: this.newValue
      },
      on: {
        click: this.toggleModal,
        closeModal: () => {
          this.closeModal();
        },
        'update:value': (value: EditableLinkProps) => {
          this.newValue = value;
        }
      }
    });

    return createElement(PopupModal, {
      scopedSlots: {
        reference: () => {
          return link;
        },
        default: () => {
          return linkToolbar;
        }
      },
      props: {
        show: this.editable && this.modalOpen,
        options: {
          placement: this.newValue.popupPlacement || 'right-start'
        }
      },
      directives: [
        {
          name: 'click-outside',
          value: { handler: this.closeModal }
        }
      ]
    });
  }
}
</script>
