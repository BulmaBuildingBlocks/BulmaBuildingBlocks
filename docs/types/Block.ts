import { VueConstructor } from 'vue';

export interface Block {
  title: string;
  component: VueConstructor;
  content?: any;
  color?: string;
  container?: boolean;
}

export interface BlockListObject {
  [x: string]: VueConstructor;
}
