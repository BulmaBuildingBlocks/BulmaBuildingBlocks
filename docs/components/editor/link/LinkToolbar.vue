<template>
  <div class="image-toolbar card">
    <header class="card-header">
      <p class="card-header-title">
        Configuration
      </p>
      <a class="card-header-icon" aria-label="close" @click="closeModal">
        <b-icon pack="fa" icon="times" />
      </a>
    </header>
    <div class="card-content">
      <b-field label="Label" placeholder="http://">
        <b-input v-model="newValue.label" />
      </b-field>
      <b-field label="Link">
        <b-input v-model="newValue.href" />
      </b-field>
      <b-field label="Type">
        <b-select v-model="newValue.linkType">
          <option v-for="(option, key) in LinkType" :key="key" :value="option">
            {{ key }}
          </option>
        </b-select>
      </b-field>

      <b-field label="Color">
        <color-picker
          :color="newValue.status"
          @color="newValue.status = $event"
        />
      </b-field>
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, PropSync } from 'nuxt-property-decorator';
import { EditableLinkProps, LinkType } from '~/components/editor/link/types';
import ColorPicker from '~/components/ColorPicker.vue';

/***
 * Editing Features:
 * Upload and replace image
 * Alt tag
 * Link
 */

@Component({
  components: {
    ColorPicker
  }
})
export default class LinkToolbar extends Vue {
  @PropSync('value', { type: Object }) newValue!: EditableLinkProps;

  LinkType = LinkType;

  closeModal() {
    this.$emit('closeModal');
  }
}
</script>
