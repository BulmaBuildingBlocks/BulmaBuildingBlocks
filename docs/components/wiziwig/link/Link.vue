<script lang="ts">
import { CreateElement } from 'vue';
import { Vue, Component, Prop, PropSync } from 'nuxt-property-decorator';
import ButtonToolbar from './LinkToolbar.vue';
import PopupModal from '~/components/global/PopupModal.vue';

import { EditableLinkProps } from '~/components/wiziwig/link/types';
import ClickOutside from '~/directives/click-outside';

/***
 * Editting Features:
 * Upload and replace button
 * Alt tag
 * Link
 */

@Component({
  directives: {
    ClickOutside
  }
})
export default class Link extends Vue {
  @Prop(Boolean) editable!: boolean;
  @PropSync('value', { type: Object }) newValue!: EditableLinkProps;

  modalOpen = false;

  toggleModal(event: Event) {
    event.preventDefault();

    this.modalOpen = !this.modalOpen;
  }

  openModal(event: Event) {
    event.preventDefault();

    this.modalOpen = true;
  }

  closeModal() {
    this.modalOpen = false;
  }

  render(createElement: CreateElement) {
    const button = createElement(
      'a',
      {
        class: this.newValue.styles,
        on: {
          click: this.toggleModal
        },
        attrs: {
          href: this.newValue.href
        }
      },
      this.newValue.label
    );

    const buttonToolbar = createElement(ButtonToolbar, {
      props: {
        value: this.newValue
      },
      on: {
        click: this.toggleModal,
        closeModal: () => {
          this.closeModal();
        },
        'update:value': (value: EditableLinkProps) => {
          this.newValue = value;
        }
      }
    });

    const popupModal = createElement(PopupModal, {
      scopedSlots: {
        reference: () => {
          return button;
        },
        default: () => {
          return buttonToolbar;
        }
      },
      props: {
        show: this.editable && this.modalOpen,
        options: {
          placement: this.newValue.popupPlacement || 'right-start'
        }
      },
      directives: [
        {
          name: 'click-outside',
          value: { handler: this.closeModal }
        }
      ]
    });

    return popupModal;
  }
}
</script>
