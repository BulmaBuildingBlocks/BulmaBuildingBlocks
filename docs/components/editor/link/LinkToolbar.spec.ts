import { shallowMount, Wrapper } from '@vue/test-utils';
import LinkToolbar from '~/components/editor/link/LinkToolbar.vue';
import { EditableLinkProps, LinkType } from '~/components/editor/link/types';

describe('LinkToolbar.vue', () => {
  let wrapper: Wrapper<LinkToolbar & { [key: string]: any }>;

  beforeEach(() => {
    // Mount the component
    const props: EditableLinkProps = {
      label: '',
      linkType: LinkType.Link
    };

    wrapper = shallowMount(LinkToolbar, {
      propsData: { value: props }
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('has the expected html structure', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance).toBeTruthy();
  });

  // Inspect the raw component options
  test('has a close modal function', () => {
    expect(typeof wrapper.vm.closeModal).toBe('function');
  });

  // Emit event from close modal
  test('has a close modal function', async () => {
    // call function to emit event
    wrapper.vm.closeModal();

    await wrapper.vm.$nextTick(); // Wait until $emits have been handled

    expect(wrapper.emitted().closeModal).toBeTruthy();
  });
});
