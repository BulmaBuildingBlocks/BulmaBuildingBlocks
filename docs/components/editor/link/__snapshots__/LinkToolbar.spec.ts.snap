// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkToolbar.vue has the expected html structure 1`] = `
<div
  class="image-toolbar card"
>
  <header
    class="card-header"
  >
    <p
      class="card-header-title"
    >
      
      Configuration
    
    </p>
     
    <a
      aria-label="close"
      class="card-header-icon"
    >
      <b-icon-stub
        icon="times"
        pack="fa"
      />
    </a>
  </header>
   
  <div
    class="card-content"
  >
    <b-field-stub
      addons="true"
      label="Label"
      placeholder="http://"
    >
      <b-input-stub
        customclass=""
        hascounter="true"
        type="text"
        usehtml5validation="true"
        value=""
      />
    </b-field-stub>
     
    <b-field-stub
      addons="true"
      label="Link"
    >
      <b-input-stub
        customclass=""
        hascounter="true"
        type="text"
        usehtml5validation="true"
      />
    </b-field-stub>
     
    <b-field-stub
      addons="true"
      label="Type"
    >
      <b-select-stub
        usehtml5validation="true"
        value="Link"
      >
        <option
          value="Link"
        >
          
          Link
        
        </option>
        <option
          value="button"
        >
          
          Button
        
        </option>
        <option
          value="button is-rounded"
        >
          
          Rounded
        
        </option>
        <option
          value="button is-outlined"
        >
          
          Outlined
        
        </option>
        <option
          value="button is-rounded is-outlined"
        >
          
          Rounded Outlined
        
        </option>
      </b-select-stub>
    </b-field-stub>
     
    <b-field-stub
      addons="true"
      label="Color"
    >
      <color-picker-stub
        color=""
      />
    </b-field-stub>
  </div>
</div>
`;
